pandas/core/indexers.py

```diff
@@ -164,7 +164,7 @@ def check_setitem_lengths(indexer, value, values) -> bool:
         #  a) not necessarily 1-D indexers, e.g. tuple
         #  b) boolean indexers e.g. BoolArray
         if is_list_like(value):
-            if len(indexer) != len(value):
+            if len(indexer) != len(value) and values.ndim == 1:
```

Comment:
would the ndim check here be necessary if EAs all supported 2D?

Reply:
Yes, we are passing here through with something like
```
df = pd.DataFrame(data={
    'col1': [1, 2, 3, 4],
    'col2': [3, 4, 5, 6],
    'col3': [6, 7, 8, 9],
})
rhs = np.array([1, 2, 3])
df.iloc[[1]] = rhs
```

too. This should set row-wise, but the check assumes it sets column-wise